tests/test_extract_contained_procedures.py

```diff
@@ -0,0 +1,502 @@
+    """
+    fcode = """
+        subroutine outer()
+            implicit none
+            integer :: x
+            x = 42
+            call inner()
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                z = x + y
+            end subroutine inner
+            function f() result(res)
+                integer :: y
+                integer :: z
+                integer :: res
+                y = 1
+                z = y
+                res = 2 * z
+            end function f
+        end subroutine outer 
+    """
+    src = Sourcefile.from_source(fcode)
+    outer = src.routines[0]
+    extract_contained_procedures(outer)
+    # NOTE: Functions in Loki are also typed as Subroutines.
+    assert not any(isinstance(r, Subroutine) for r in outer.contains.body)
+
+def test_extract_contained_procedures_basic_array():
+    """
+    Tests that a global array variable (and a scalar) is correctly added as argument of `inner`.
+    """
+
+    fcode = """
+        subroutine outer()
+            implicit none
+            integer :: x
+            real :: arr(3)
+            arr = 71.0
+            x = 42
+            call inner()
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                z = x + y + arr(1)
+            end subroutine inner
+        end subroutine outer 
+    """
+    src = Sourcefile.from_source(fcode)
+    routines = extract_contained_procedures(src.routines[0])
+    assert len(routines) == 1
+    inner = routines[0]
+    outer = src.routines[0]
+    assert 'x' in (var.name for var in FindVariables().visit(inner.spec))
+    assert 'x' in (var.name for var in inner.arguments)
+    assert 'arr' in (var.name for var in FindVariables().visit(inner.spec))
+    assert 'arr' in (var.name for var in inner.arguments)
+
+    call = FindNodes(CallStatement).visit(outer.body)[0]
+    assert 'x' in (var.name for var in call.arguments)
+    assert 'arr' in (var.name for var in call.arguments)
+
+def test_extract_contained_procedures_basic_import():
+    """
+    Tests that a global imported binding is correctly introduced to the contained subroutine.
+    """
+
+    fcode = """
+        subroutine outer()
+            use constants, only: c1, c2
+            implicit none
+            integer :: x
+            x = 42 + c1
+            call inner()
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                z = x + y + c2
+            end subroutine inner
+        end subroutine outer 
+    """
+    src = Sourcefile.from_source(fcode)
+    routines = extract_contained_procedures(src.routines[0])
+    assert len(routines) == 1
+    inner = routines[0]
+    assert "c2" in inner.import_map.keys()
+    assert not "c1" in inner.import_map.keys()
```

Comment:
'.keys()' is redundant for checking if a key exists in a dict

Reply:
Thanks, indeed it is. I removed all instances of this.